### PR TITLE
feat: doc improvement LA respondent category (FPLA-2770)

### PR DESCRIPTION
**Pull request by Jamesnelson117**
**Before creating a pull request make sure that:**

- [x] PR's title includes JIRA ticket number and follow [Conventional Commits](https://www.conventionalcommits.org/) specification
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-2770

### Change description ###
Respondent statements can now be added via manage documents event (LA only)

https://user-images.githubusercontent.com/4429628/112662526-b1b63100-8e4f-11eb-89fd-06014dcd45e3.mov

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/document/ManageDocumentService.java [hunk 1]
```diff
@@ -278,6 +281,58 @@
         return setDateTimeUploadedOnSupportingEvidence(currentSupportingDocuments, previousSupportingDocuments);
     }
 
+    public List<Element<SupportingEvidenceBundle>> getRespondentStatementFurtherEvidenceCollection(CaseData caseData,
+                                                                                                   UUID id) {
+        Optional<Element<RespondentStatement>> respondentStatement = caseData.getRespondentStatementByRespondentId(id);
```
Comment (tomaszpowroznik) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/document/ManageDocumentService.java:286]: can be written as:
       return caseData.getRespondentStatementByRespondentId(id)
            .map(Element::getValue)
            .map(RespondentStatement::getSupportingEvidenceBundle)
            .orElse(defaultSupportingEvidences());

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/document/ManageDocumentService.java [hunk 2]
```diff
@@ -333,4 +388,14 @@ private void sortByDateUploaded(List<Element<SupportingEvidenceBundle>> evidence
             return date1.compareTo(date2);
         });
     }
+
+    private String getRespondentFullName(CaseData caseData, UUID respondentId) {
+        Optional<Element<Respondent>> optionalRespondentElement = caseData.getRespondentByUUID(respondentId);
```
Comment (tomaszpowroznik) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/document/ManageDocumentService.java:393]: minor: can be:
```
        return caseData.getRespondentByUUID(respondentId)
            .map(Element::getValue)
            .map(Respondent::getParty)
            .map(RespondentParty::getFullName)
            .orElseThrow(()->new RespondentNotFoundException(respondentId));

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/document/ManageDocumentService.java [hunk 3]
```diff
@@ -278,6 +281,58 @@
...
 
+    public List<Element<SupportingEvidenceBundle>> getRespondentStatementFurtherEvidenceCollection(CaseData caseData,
+                                                                                                   UUID id) {
+        Optional<Element<RespondentStatement>> respondentStatement = caseData.getRespondentStatementByRespondentId(id);
+
+        return respondentStatement.isPresent() ? respondentStatement.get().getValue().getSupportingEvidenceBundle()
+            : defaultSupportingEvidences();
+    }
+
+    public List<Element<RespondentStatement>> getUpdatedRespondentStatements(CaseData caseData) {
```
Comment (tomaszpowroznik) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/document/ManageDocumentService.java:292]: maybe this is a bit simpler?
```
    public List<Element<RespondentStatement>> getUpdatedRespondentStatements(CaseData caseData) {
        List<Element<RespondentStatement>> respondentStatementDocuments = caseData.getRespondentStatements();
        UUID selectedRespondentId = getSelectedRespondentId(caseData);
        String respondentFullName = getRespondentFullName(caseData, selectedRespondentId);
        List<Element<SupportingEvidenceBundle>> newBundle = caseData.getSupportingEvidenceDocumentsTemp();
        
        Element<RespondentStatement> respondentStatement
            = caseData.getRespondentStatementByRespondentId(selectedRespondentId)
            .orElse(element(RespondentStatement.builder()
                .respondentId(selectedRespondentId)
                .respondentName(respondentFullName)
                .build()));
        
        if(newBundle.isEmpty()){
            respondentStatementDocuments.remove(respondentStatement);
        } else {
            respondentStatement.getValue().setSupportingEvidenceBundle(newBundle);    
        }
        
        return respondentStatementDocuments;
    }

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/document/ManageDocumentService.java [hunk 4]
```diff
@@ -278,6 +281,58 @@
         return setDateTimeUploadedOnSupportingEvidence(currentSupportingDocuments, previousSupportingDocuments);
     }
 
+    public List<Element<SupportingEvidenceBundle>> getRespondentStatementFurtherEvidenceCollection(CaseData caseData,
```
Comment (tomaszpowroznik) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/document/ManageDocumentService.java:284]: should it not go to ManageDocumentServiceLA if functionality exists only for LA

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/CaseDataTest.java [hunk 5]
```diff
@@ -1384,6 +1385,125 @@ void shouldReturnAnEmptyOptionalWhenExpectedOrderIsNotContainedWithinHearingOrde
...
+    @Nested
+    class BuildRespondentStatementDynamicList {
+        @Test
+        void shouldBuildDynamicRespondentStatementListFromRespondents() {
+            List<Element<Respondent>> respondents = createRespondents();
+
+            CaseData caseData = CaseData.builder().respondents1(respondents).build();
+            IncrementalInteger i = new IncrementalInteger(1);
+
+            DynamicList expectedDynamicList = ElementUtils
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/model/CaseDataTest.java:1397]: minor: not big fan of this, as test code uses same logic as production code so wont be able to spot issues, there is DynamicListHelper class that may be used

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/CaseDataTest.java [hunk 6]
```diff
@@ -1384,6 +1385,125 @@ void shouldReturnAnEmptyOptionalWhenExpectedOrderIsNotContainedWithinHearingOrde
...
+
+            Optional<Element<RespondentStatement>> optionalRespondentStatementElement
+                = caseData.getRespondentStatementByRespondentId(elementId);
+
+            assertThat(optionalRespondentStatementElement).isNotPresent();
+        }
+    }
+
+    @Nested
+    class GetRespondentByUUID {
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/model/CaseDataTest.java:1476]: getRespondentById, getRespondentByUUID sounds like getRespondentByString ot getRespondentByLong

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/ManageDocumentServiceTest.java [hunk 7]
```diff
@@ -815,6 +821,267 @@ void shouldNotRemoveHearingFurtherEvidenceBundleElementWhenDocumentsForThatHeari
...
+                    element(RespondentStatement.builder()
+                        .respondentId(UUID.randomUUID())
+                        .supportingEvidenceBundle(supportingEvidenceBundleTwo)
+                        .build())))
+                .build();
+
+            List<Element<SupportingEvidenceBundle>> actualBundle
+                = manageDocumentService.getRespondentStatementFurtherEvidenceCollection(caseData, selectedRespondentId);
+
+            assertThat(actualBundle.get(0).getValue()).isEqualTo(SupportingEvidenceBundle.builder().build());
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/ManageDocumentServiceTest.java:871]: This one is stronger assertion as it check that array has only one element:
``` assertThat(actualBundle).extracting(Element::getValue).containsExactly(SupportingEvidenceBundle.builder().build());

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/ManageDocumentServiceTest.java [hunk 8]
```diff
@@ -815,6 +821,267 @@ void shouldNotRemoveHearingFurtherEvidenceBundleElementWhenDocumentsForThatHeari
...
+        }
+
+        @Test
+        void shouldReturnEmptySupportingEvidenceDocumentsWhenCaseDoesNotContainRespondentStatements() {
+            CaseData caseData = CaseData.builder().build();
+
+            List<Element<SupportingEvidenceBundle>> actualBundle
+                = manageDocumentService.getRespondentStatementFurtherEvidenceCollection(caseData, selectedRespondentId);
+
+            assertThat(actualBundle.get(0).getValue()).isEqualTo(SupportingEvidenceBundle.builder().build());
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/ManageDocumentServiceTest.java:881]: same as above

File: service/src/main/java/uk/gov/hmcts/reform/fpl/exceptions/RespondentNotFoundException.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package uk.gov.hmcts.reform.fpl.exceptions;
+
+import java.util.UUID;
+
+public class RespondentNotFoundException extends IllegalStateException {
+    public RespondentNotFoundException() {
+        super("Respondent not found");
+    }
+
+    public RespondentNotFoundException(UUID hearingBookingId) {
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/exceptions/RespondentNotFoundException.java:10]: hearingBookingId -> respondentId

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/ManageDocumentServiceTest.java [hunk 10]
```diff
@@ -815,6 +821,267 @@ void shouldNotRemoveHearingFurtherEvidenceBundleElementWhenDocumentsForThatHeari
...
+                            element(supportingEvidenceBundleId, SupportingEvidenceBundle.builder().build())
+                        ))
+                        .respondentName("David Stevenson")
+                        .build())))
+                .build();
+
+            List<Element<RespondentStatement>> updatedRespondentStatements =
+                manageDocumentService.getUpdatedRespondentStatements(caseData);
+
+            assertThat(updatedRespondentStatements.size()).isEqualTo(1);
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/ManageDocumentServiceTest.java:921]: size assertion is not needed as next assertion covers it, best would be to use:
```
 assertThat(updatedRespondentStatements).containsExactly(
                element(...
 ```
 
 it gives better error on failure

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java [hunk 11]
```diff
@@ -147,6 +147,18 @@
     @NotEmpty(message = "Add the respondents' details")
     private final List<@NotNull(message = "Add the respondents' details") Element<Respondent>> respondents1;
 
+    public DynamicList buildRespondentStatementDynamicList(UUID selected) {
+        IncrementalInteger i = new IncrementalInteger(1);
+
+        return asDynamicList(getAllRespondents(), selected,
+            respondent -> String.format("Respondent %d statements - %s",
```
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java:154]: Agreed with Collette that it should just read respondent name.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/document/ManageDocumentService.java (hunk 4 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/document/ManageDocumentService.java:284]: Next story is to implement the same feature only for the admin. Hence why I placed this functionality here. Let's keep here.

File: ccd-definition/CaseField/CareSupervision/manageDocuments.json [hunk 12]
```diff
@@ -91,5 +91,14 @@
     "Label": "## C2 supporting documents",
     "SecurityClassification": "Public",
     "Searchable": "N"
+  },
+  {
```
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply ccd-definition/CaseField/CareSupervision/manageDocuments.json:95]: Next story is to implement the same functionality for admin etc. Let's keep here.

File: ccd-definition/AuthorisationCaseField/CareSupervision/la-managing.json [hunk 13]
```diff
@@ -846,6 +846,13 @@
     "UserRole": "[LAMANAGING]",
     "CRUD": "R"
   },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "respondentStatements",
+    "UserRole": "[LAMANAGING]",
+    "CRUD": "R"
```
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply ccd-definition/AuthorisationCaseField/CareSupervision/la-managing.json:854]: Yeah good spot, `hearingFurtherEvidenceDocuments` should be CRUD also. Let's confirm and fix in another story however, will update permissions above to be CRUD.

File: ccd-definition/AuthorisationCaseField/CareSupervision/court-admin.json [hunk 14]
```diff
@@ -1308,6 +1308,13 @@
     "UserRole": "caseworker-publiclaw-courtadmin",
     "CRUD": "CRUD"
   },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "respondentStatements",
+    "UserRole": "caseworker-publiclaw-courtadmin",
+    "CRUD": "CRUD"
```
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply ccd-definition/AuthorisationCaseField/CareSupervision/court-admin.json:1316]: Agreed to leave these permissions as they where as they are being overwritten in `AuthorisationComplexType`

File: ccd-definition/AuthorisationCaseField/CareSupervision/judiciary.json [hunk 15]
```diff
@@ -937,6 +937,13 @@
     "UserRole": "caseworker-publiclaw-judiciary",
     "CRUD": "CRUD"
   },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "respondentStatements",
+    "UserRole": "caseworker-publiclaw-judiciary",
+    "CRUD": "CRUD"
```
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply ccd-definition/AuthorisationCaseField/CareSupervision/judiciary.json:945]: Agreed to leave these permissions as they where as they are being overwritten in AuthorisationComplexType

**Review by tomaszpowroznik — APPROVED** [review]

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/CaseDataTest.java [hunk 16]
```diff
@@ -1384,6 +1385,120 @@ void shouldReturnAnEmptyOptionalWhenExpectedOrderIsNotContainedWithinHearingOrde
...
+                respondentStatementElementOne,
+                element(RespondentStatement.builder().build()));
+
+            CaseData caseData = CaseData.builder().respondentStatements(respondentStatements).build();
+
+            Optional<Element<RespondentStatement>> optionalRespondentStatementElement
+                = caseData.getRespondentStatementByRespondentId(elementId);
+
+            assertThat(optionalRespondentStatementElement).isPresent();
+            assertThat(optionalRespondentStatementElement).contains(respondentStatementElementOne);
```
Comment (joez-f) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/model/CaseDataTest.java:1452]: codesmell: can be chained:

assertThat(optionalRespondentStatementElement)
.isPresent()
.contains(respondentStatementElementOne);

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/CaseDataTest.java [hunk 17]
```diff
@@ -1384,6 +1385,120 @@ void shouldReturnAnEmptyOptionalWhenExpectedOrderIsNotContainedWithinHearingOrde
...
+
+            List<Element<Respondent>> respondents = List.of(
+                respondentOneElement,
+                element(Respondent.builder().build()));
+
+            CaseData caseData = CaseData.builder().respondents1(respondents).build();
+            Optional<Element<Respondent>> optionalRespondentElement = caseData.findRespondent(elementId);
+
+            assertThat(optionalRespondentElement).isPresent();
+            assertThat(optionalRespondentElement).contains(respondentOneElement);
```
Comment (joez-f) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/model/CaseDataTest.java:1486]: can be chained

**Review by nraghu32 — APPROVED** [review]

**Pull request merged**